### ISSUE TITLE
Names of inputs to derivative functions

### ISSUE THREAD

**Issue by modelica-trac-importer** [modelica/ModelicaSpecification#1548]
**Reported by stefanv on 29 Jul 2014 16:16 UTC**
Consider the function F:

```mo
function F
    input Real x;
    input Real y;
    output Real z;
algorithm
    z := ...;
    annotation (derivative(zeroDerivative=y) = F_der);
end F;
```

In the function _F_der_, the derivative of _x_ can have any name the user chooses, be it _dx_, _der_x_, _slope_of_x_, or whatever.

But, does _x_ itself have to have the same name?

Does _y_? If not, does the zeroDerivative annotation in _F_ have to specify the name of _y_ as it appears in _F_ (which is of course _y_), or the name that _y_ has in _F_der_?

In short, which, if any, of the implementations of _F_der_ below are legal given _F_ above?

```mo
function F_der
    input Real xWithAnotherName;
    input Real y;
    input Real x_der;
    output Real z_der;
algorithm
    z_der := ...;
end F;
```

```mo
function F_der
    input Real x;
    input Real yWithAnotherName;
    input Real x_der;
    output Real z_der;
algorithm
    z_der := ...;
end F;
```

----

Migrated-From: https://trac.modelica.org/Modelica/ticket/1548

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 1 Dec 2014 15:44 UTC**
Language group:
Derivative functions must preserve all existing arguments (with the same type, same name in the same order) for inputs and outputs, i.e. otherwise it is an error.

Meta-comment: open another ticket to consistently use "must" for strict rules and "should" for recommendations that don't need to be enforced.

**Comment by modelica-trac-importer** [comment]
**Modified by hansolsson on 10 Dec 2014 15:40 UTC**

**Comment by modelica-trac-importer** [comment]
**Comment by jmattsson on 10 Dec 2014 15:46 UTC**
Decision was made at design meeting. Now need to update spec.

**Comment by modelica-trac-importer** [comment]
**Comment by stefanv on 22 Jan 2015 17:13 UTC**
A further clarification: Do inputs that are declared as `noDerivative` need to have matching names? If so, `Modelica.Media.Water.IF97_Utilities.T_ph_der` is incorrect. I have filed a separate ticket, #1650, to report that bug.

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 17 Jul 2015 08:19 UTC**
Replying to [comment:1 hansolsson]:
> Language group:
> Derivative functions must preserve all existing arguments (with the same type, same name in the same order) for inputs and outputs, i.e. otherwise it is an error.
> 
> Meta-comment: open another ticket to consistently use "must" for strict rules and "should" for recommendations that don't need to be enforced.

The input-part is clear, but the output-part is less clear.

Basically the derivative does not have the same outputs as the function (it has the derivative of the outputs) and thus the restrictions do not make sense. Thus I assume above it is meant that the outputs have the same type and order, but ignoring the name-part - this is also consistent with the example in the ticket-body (where the function has 'z' as output and the derivative 'z_der').

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 17 Jul 2015 08:35 UTC**
Resolved in r8353

**Comment by modelica-trac-importer** [comment]
**Modified by hansolsson on 17 Jul 2015 11:23 UTC**

**Comment by modelica-trac-importer** [comment]
**Comment by stefanv on 17 Jul 2015 14:18 UTC**
I agree with the interpretation in comment 5 above.
